### PR TITLE
✨  log out the user when not active

### PR DESCRIPTION
**Pull request by shkurata**
Ref: #293

### CODE REVIEW

File: api/src/schema/types/auth.ts [hunk 1]
```diff
@@ -84,7 +84,9 @@ builder.objectType(JWT, {
       nullable: true,
       resolve: (jwt) => {
         const expiredAt = dateFromJwtTimestamp(jwt.exp);
-        return expiredAt == null ? null : ms(expiredAt.getTime() - Date.now());
+        return expiredAt == null
+          ? null
+          : (expiredAt.getTime() - Date.now()).toString();
```
Comment (moeriki) [inline api/src/schema/types/auth.ts:89]: You should update the field type to number instead of casting to string.

File: app/src/auth.ts [hunk 2]
```diff
@@ -50,10 +56,24 @@ export const useAuthStore = defineStore('auth', () => {
     }
   }
 
+  function setSessionExpirationTimer(authData: Auth | undefined): void {
+    const userData = authData?.currentUser;
+    const EXTRA_DELAY = 1000;
```
Comment (moeriki) [inline app/src/auth.ts:61]: Why add an extra delay? I would rather do the opposite. Refresh faster.

File: app/src/auth.ts [hunk 3]
```diff
@@ -50,10 +56,24 @@ export const useAuthStore = defineStore('auth', () => {
...
+        }, expiresIn + EXTRA_DELAY);
+      }
+    }
+  }
+
   return {
     name: computed(() => auth.value.currentUser?.name),
     isLoggedIn: computed(() => auth.value.currentUser != null),
     refresh,
+    setSessionExpirationTimer,
```
Comment (moeriki) [inline app/src/auth.ts:76]: Why expose function?

File: app/src/auth.ts [hunk 4]
```diff
@@ -50,10 +56,24 @@ export const useAuthStore = defineStore('auth', () => {
     }
   }
 
+  function setSessionExpirationTimer(authData: Auth | undefined): void {
+    const userData = authData?.currentUser;
+    const EXTRA_DELAY = 1000;
+    if (userData) {
+      const expiresIn = Number(userData.tokenSet.refreshToken.expiresIn);
+      if (!Number.isNaN(expiresIn)) {
+        setTimeout(() => {
```
Comment (moeriki) [inline app/src/auth.ts:65]: The previous timer should be canceled.

File: app/src/auth.ts [hunk 5]
```diff
@@ -50,10 +56,24 @@ export const useAuthStore = defineStore('auth', () => {
     }
   }
 
+  function setSessionExpirationTimer(authData: Auth | undefined): void {
```
Comment (moeriki) [inline app/src/auth.ts:59]: The argument is useless as the variable is available in scope. Also why make a function? This could be inline.

**Review by moeriki — CHANGES_REQUESTED** [review]

File: app/src/auth.ts [hunk 6]
```diff
@@ -34,12 +39,27 @@ export const useAuthStore = defineStore('auth', () => {
     currentUser: null,
   });
 
+  const timer = ref<NodeJS.Timeout | undefined>();
```
Comment (moeriki) [inline app/src/auth.ts:42]: That's a strange type for a browser environment. Is number possible?

File: app/src/auth.ts [hunk 7]
```diff
@@ -34,12 +39,27 @@ export const useAuthStore = defineStore('auth', () => {
...
     try {
       const authQuery = await apolloClient.query({
         query: AuthCurrentUserDocument,
       });
       auth.value = authQuery.data.auth;
+
+      const { currentUser } = auth.value;
+
+      if (currentUser) {
+        const expiresIn = Number(currentUser.tokenSet.refreshToken);
```
Comment (moeriki) [inline app/src/auth.ts:54]: You cast a tokenset to a number? That's probably wrong.

**Review by moeriki — CHANGES_REQUESTED** [review]

File: app/src/auth.ts [hunk 8]
```diff
@@ -34,12 +39,27 @@ export const useAuthStore = defineStore('auth', () => {
...
       const authQuery = await apolloClient.query({
         query: AuthCurrentUserDocument,
       });
       auth.value = authQuery.data.auth;
+
+      const { currentUser } = auth.value;
+
+      if (currentUser) {
+        const expiresIn = Number(currentUser.tokenSet.refreshToken.expiresIn);
+        if (!Number.isNaN(expiresIn)) {
```
Comment (moeriki) [inline app/src/auth.ts:55]: Isn't `currentUser.tokenSet.refreshToken.expiresIn` already a number? I think the cast and `NaN` check is not necessary.

File: app/src/auth.ts [hunk 9]
```diff
@@ -34,12 +39,27 @@ export const useAuthStore = defineStore('auth', () => {
...
+          if (timer.value != null) {
+            clearTimeout(timer.value);
+          }
```
Comment (moeriki) [inline app/src/auth.ts:58]: `if` condition isn't necessary. `clearTimeout` can take `number | undefined` anyways.

File: app/src/auth.ts [hunk 10]
```diff
@@ -34,12 +39,27 @@ export const useAuthStore = defineStore('auth', () => {
...
+          // eslint-disable-next-line @typescript-eslint/no-misused-promises
+          timer.value = window.setTimeout(refresh, expiresIn);
```
Comment (moeriki) [inline app/src/auth.ts:60]: The ESLint disable isn't really necessary. Better to do `() => void refresh()`.

`void` is a convenient way to communicate in code that a function result can be discarded.

**Review by moeriki — CHANGES_REQUESTED** [review]
There's a bunch of change in the generated files. Make sure to run `yarn build` and amend the changes again to make sure it's correct.

**Comment by shkurata** [comment]
> `if` condition isn't necessary. `clearTimeout` can take `number | undefined` anyways. 

I had to leave the `null` check because I get errors.

**Review by moeriki — APPROVED** [review]

**Pull request merged**